<a href="https://colab.research.google.com/github/GOPIKAV77/Python/blob/main/MLE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

MAXIMUM LIKELIHOOD ESTIMATE

In [1]:
import numpy as np
from scipy.optimize import minimize
import scipy.stats as stats

In [3]:
import pandas as pd
data = {'Id':['1','2','3','4','5'],'Y':['2','6','7','9','15'],'X':['1','4','5','6','9']}
df = pd.DataFrame(data)
df

Id   Y  X
0  1   2  1
1  2   6  4
2  3   7  5
3  4   9  6
4  5  15  9

OLS SUMMARY

In [11]:
df['Y'] = pd.to_numeric(df['Y'])
df["X"] = pd.to_numeric(df["X"])

In [12]:
import statsmodels.formula.api as s
import statsmodels.api as sm
x = df['X']
y = df['Y']
x2 = sm.add_constant(x)
m = s.ols('Y~X',data = df)
m = m.fit()
m.summary()

/usr/local/lib/python3.11/dist-packages/statsmodels/stats/stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.980
Model:                            OLS   Adj. R-squared:                  0.973
Method:                 Least Squares   F-statistic:                     145.9
Date:                Mon, 24 Feb 2025   Prob (F-statistic):            0.00122
Time:                        12:48:09   Log-Likelihood:                -4.5811
No. Observations:                   5   AIC:                             13.16
Df Residuals:                       3   BIC:                             12.38
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     -0.2882      0.755     -0.382      0.728      -2.692       2.115
X              1.6176      0.134     12.079      0.001       1.191       2.044
==============================================================================
Omnibus:                          nan   Durbin-Watson:                   1.405
Prob(Omnibus):                    nan   Jarque-Bera (JB):                0.551
Skew:                           0.089   Prob(JB):                        0.759
Kurtosis:                       1.384   Cond. No.                         12.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**RESULT**

```

coef	std err	t	P>|t|	[0.025	0.975]
Intercept	-0.2882	0.755	-0.382	0.728	-2.692	2.115
X	1.6176	0.134	12.079	0.001	1.191	2.044


Y = -0.2882 + 0.134 x1

PARAMETER ESTIMATION BY MLE

In [14]:
e = m.resid

In [15]:
e

0    0.670588
1   -0.182353
2   -0.800000
3   -0.417647
4    0.729412
dtype: float64

In [16]:
np.std(e)

0.6048820983804829

In [17]:
import numpy as np
from scipy.optimize import minimize
import matplotlib.pyplot as plt

def lik(parameters):
  m = parameters[0]
  b = parameters[1]
  sigma = parameters[2]

  for i in np.arange(0, len(x)):
    y1 = m*x + b
  L = (len(x)/2*np.log(2*np.pi)*len(x)/2*np.log(sigma**2)+1/(2*sigma**2)*sum((y-y1)**2))
  return L

x = np.array([1,4,5,6,9])
y = np.array([2,6,7,9,15])
model = minimize(lik, np.array([2,2,2]), method = 'L-BFGS-B')

In [18]:
model

  message: CONVERGENCE: REL_REDUCTION_OF_F_<=_FACTR*EPSMCH
  success: True
   status: 0
      fun: -17.578665889890743
        x: [ 1.618e+00 -2.882e-01 -2.822e-01]
      nit: 19
      jac: [ 2.540e-04  5.400e-05  2.640e-04]
     nfev: 112
     njev: 28
 hess_inv: <3x3 LbfgsInvHessProduct with dtype=float64>